mqtt.go

```diff
@@ -499,6 +516,124 @@ func (mc *mqttConn) writeBulkData(b *DeviceSyncedBulkData) error {
 	return errors.New("did not receive ack packet until timeout and bulk data dropped")
 }
 
+func (mc *mqttConn) sendBulkDataRequest(b *DeviceBulkDataRequest) error {
+	var qos byte
+
+	switch b.qos {
+	case QOSAtMostOnce:
+		qos = packet.QOSAtMostOnce
+	case QOSAtLeastOnce:
+		qos = packet.QOSAtLeastOnce
+	default:
+		return errors.New("unsupported qos level")
+	}
+
+	payload, _ := json.Marshal(b)
+
+	p := packet.NewPublishPacket()
+	p.PacketID = mc.getPacketID()
+	p.Message = packet.Message{
+		Topic:   fmt.Sprintf("/devices/%d/bulkData/%s/request", mc.dc.DeviceID, b.StreamID),
+		QOS:     qos,
+		Payload: payload,
+	}
+
+	mc.outPacket <- p
+
+	logInfo("[MQTT] bulkData request delivery for packet ID %d and requestID %s", p.PacketID, b.RequestID)
+
+	if b.qos == QOSAtMostOnce {
+		return nil
+	}
+
+	logInfo("[MQTT] waiting for PUBACK for packet ID %d", p.PacketID)
+
+	select {
+	case <-time.After(deviceEventRetryInterval):
```

Comment:
I think you should just make a function to take bytes and the topic, and maybe the format string for the error. Then sendEvent and sendBulkDataRequest will just call that.

Reply:
thanks. fixed. https://github.com/moderepo/device-sdk-go/pull/28/commits/29b66bb624b63f877f0626a66843085f2378399a